var/spack/repos/builtin/packages/py-pyfr/package.py

```diff
@@ -52,7 +56,21 @@ class PyPyfr(PythonPackage):
 
     # Optional dependencies
     depends_on("py-scipy", when="+scipy", type=("build", "run"))
+    depends_on("metis@5.0:", when="+metis", type=("run"))
     depends_on("scotch@6.0:", when="+scotch", type=("run"))
-    depends_on("cuda@8.0:", when="+cuda", type=("run"))
-    depends_on("rocblas@4.5.0:", when="+hip", type=("run"))
-    depends_on("libxsmm@1.18:+shared blas=0 +large_jit_buffer", when="+libxsmm", type=("run"))
+    depends_on("cuda@8.0: +allow-unsupported-compilers", when="@:1.14.0 +cuda", type=("run"))
+    depends_on("cuda@11.4.0: +allow-unsupported-compilers", when="@1.15.0: +cuda", type=("run"))
+    depends_on("rocblas@5.2.0:", when="+hip", type=("run"))
+    depends_on("libxsmm@1.18:+shared blas=0", when="+libxsmm", type=("run"))
+
+    # Conflicts for compilers not supporting OpenMP 5.1+ from v1.15.0:
+    conflicts("%gcc@:11", when="@1.15.0: +libxsmm", msg="OpenMP 5.1+ supported compiler required!")
+
+    # Explicitly add dependencies to PYFR_LIBRARY_PATH environment variable
+    def setup_run_environment(self, env):
+        deps = ["metis", "scotch", "libxsmm", "cuda", "hip", "rocblas"]
+        pyfr_library_path = []
+        for dep in deps:
+            if "+{}".format(dep) in self.spec:
+                pyfr_library_path.extend(self.spec[dep].libs.directories)
+        env.set("PYFR_LIBRARY_PATH", ":".join(pyfr_library_path))
```

Comment:
This may also work, and doesn't require you to explicitly list which packages/variants to check:
```suggestion
        pyfr_library_path = []
        for dep in self.spec.dependencies(deptype="link"):
            pyfr_library_path.extend(self.spec[dep.name].libs.directories)
        env.set("PYFR_LIBRARY_PATH", ":".join(pyfr_library_path))
```
This will recursively list all link dependencies, not sure if there's a way to restrict it to direct dependencies.

Reply:
Thanks @adamjstewart, I considered this but feel that verbose is better in this case. 
I think that it is ready to merge.